### ISSUE TITLE
[v1.8] , [v2.0]  src/checks.c, segfault at 4 ...

### ISSUE THREAD

**Issue by qermit** [haproxy/haproxy#1337]
### Detailed Description of the Problem

haproxy randomly crashes with segfault
```
kernel: [46356.451561] haproxy18[1575]: segfault at 4 ip 000055cc105109fa sp 00007f5286f30f90 error 6 in haproxy18[55cc10475000+150000]
```

```
hex(0x000055cc105109fa - 0x55cc10475000) = 0x9b9fa
addr2line -e /usr/sbin/haproxy18 -fCi 0x9b9fa -> /usr/src/debug/haproxy-1.8.27/src/checks.c:3123
```

```
3291  out_end_tcpcheck:
3292         /* collect possible new errors */
3293         if ((conn && conn->flags & CO_FL_ERROR) || (cs && cs->flags & CS_FL_ERROR))
3294                 chk_report_conn_err(check, 0, 0);
3295 
3296         /* cleanup before leaving */
3297         check->current_step = NULL;
3298 
3299         if (check->result == CHK_RES_FAILED)
3300                 conn->flags |= CO_FL_ERROR;
```

### Expected Behavior

haproxy shouldn't crash

### Steps to Reproduce the Behavior

1) start haproxy with checks
2) wait for crash

### Do you have any idea what may have caused this?

there is no check against `conn` when  setting flag `CO_FL_ERROR`. It is checked earlier (introduced in https://github.com/haproxy/haproxy/commit/ef91c939f364367bdc48f1b234c138d743abf56c) 

### What is your configuration?

```haproxy
frontend frontend
    mode tcp
    bind 127.0.0.1:5999
    default_backend frontend

backend backend
    mode tcp
    option tcp-check
    option allbackups
    balance roundrobin
    tcp-check connect
    tcp-check send test_network_script:\ PING\r\n\r\n
    tcp-check expect string PONG
    server alpha_0 127.0.0.1:6000 maxconn 128 check inter 10s
    server beta_0 127.0.0.1:6001 maxconn 128 check inter 10s backup
```

### Output of `haproxy -vv`

```plain
HA-Proxy version 1.8.27-493ce0b 2020/11/06
Copyright 2000-2020 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_SYSTEMD=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
Running on OpenSSL version : OpenSSL 1.1.1g FIPS  21 Apr 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace
```

### Output of `uname -a`

```
Linux dev1stage0sip.v 3.10.0-1160.31.1.el7.x86_64 #1 SMP Thu Jun 10 13:32:12 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ rpm -q haproxy18
haproxy18-1.8.27-1.el7.x86_64
```

### Last Outputs and Backtraces

```plain
The machine was not used at all

journalctl -u haproxy18
...
Jul 21 13:28:36 dev1stage0sip.v haproxy18[1185]: Proxy frontend started.
Jul 21 16:26:33 dev1stage0sip.v haproxy18[1185]: [ALERT] 201/132836 (1185) : Current worker 1198 exited with code 139
Jul 21 16:26:33 dev1stage0sip.v haproxy18[1185]: [ALERT] 201/132836 (1185) : exit-on-failure: killing every workers with SIGTERM
Jul 21 16:26:33 dev1stage0sip.v haproxy18[1185]: [WARNING] 201/132836 (1185) : All workers exited. Exiting... (139)
Jul 21 16:26:33 dev1stage0sip.v systemd[1]: haproxy18.service: main process exited, code=exited, status=139/n/a
...
```

**Comment by qermit** [comment]
Bug does not exist since b2c2e0fcca09acc6e6527cb5d90a18a9674e7a78

**Comment by capflam** [comment]
Thanks for the detailed report. The bug should be fixed now, in 2.0 (d86b4bf44) and 1.8 (3d029f82f).
